closed ring segment buffer

todostreaming opened the pull request:
Have you thought of implementing a closed ring buffer for segmented protocols (Youtube, HLS, HDS...) ? I mean, if some segments are not downloaded at time, on --player-external-http, just rotate the already downloaded to be served to the player. As an option.

Thanks for you great soft.
Regards
/ts

chrippa committed:
Merge branch 'release/1.12.2' into develop

betagan committed:
fix vaughnlive plugin #897

chrippa committed:
Merge pull request #898 from betagan/develop

fix vaughnlive plugin #897

blxd committed:
fixed tvcatchup.com plugin, the website layout changed and the method to find the stream URLs needed to be updated.

chrippa committed:
Merge pull request #900 from blxd/bugfix/tvcatchup

Fix for the tvcatchup.com plugin

chrippa committed:
plugins.twitch: Handle subdomains with dash in them, e.g. en-gb.

Resolves #906.

maop committed:
Add Beam.pro plugin.

bdien committed:
Meerkat plugin

chrippa committed:
Merge pull request #922 from bdien/develop

Add Meerkat plugin

chrippa committed:
Merge pull request #920 from maop/beam-plugin

blxd committed:
tvcatchup now returns a variant playlist

blxd committed:
tvplayer.com only works with a browser user agent

chrippa committed:
Merge pull request #928 from blxd/bugfix/tvcatchup

TVCatchup returns a variant playlist

chrippa committed:
Merge pull request #929 from blxd/bugfix/tvplayer

TVPlayer.com now only works with a browser user agent header

blxd committed:
not all channels return hlsvariant playlists

Top-Ranger committed:
Added plugin for blip.tv VOD

Top-Ranger committed:
Updated blip.tv plugin

Fixed flaws according to https://github.com/chrippa/livestreamer/pull/936

wolftankk committed:
get azubu live status from api

chrippa committed:
Merge pull request #942 from wolftankk/develop

update azubutv plugin

chrippa committed:
Merge pull request #936 from Top-Ranger/bliptv

Added plugin for blip.tv VOD

wolftankk committed:
use new api get stream info

wolftankk committed:
fix video_player error

intact committed:
Added plugin for streamup.com

intact committed:
Updated douyutv.com plugin

hannespetur committed:
plugin for Ruv - the Icelandic national television - was added

hannespetur committed:
removed print statements and started to use quality key as audio if the url extensions is mp3

hannespetur committed:
the plugin added to the plugin matrix

hannespetur committed:
removed unused import

hannespetur committed:
alphabetical order is hard

hannespetur committed:
removed redundant assignments of best/worst quality

wolftankk committed:
Merge remote-tracking branch 'upstream/develop' into develop

hannespetur committed:
HLS support added for the Ruv plugin

hannespetur committed:
Ruv plugin: returning generators instead of a dict

intact committed:
plugins.streamupcom: Check live status

chrippa committed:
Merge pull request #956 from hannespetur/develop

Plugin for Rúv - The Icelandic national television

livescope committed:
Add VOD/replay support for periscope.tv

chrippa committed:
Merge pull request #959 from livescope/develop

Add VOD/replay support for periscope.tv

intact committed:
plugins.artetv: Update json regex

chrippa committed:
Merge pull request #961 from intact/artetv-fix

plugins.artetv: Update json regex

Commit:
added expressen plugin

Commit:
added expressen plugin

Commit:
update() => assign with subscript

Commit:
added entry for expressen

intact committed:
plugins.streamupcom: Update for API change

Ell committed:
added support for livecoding.tv

Ell committed:
removed printing

Ell committed:
updated plugin matrix

chrippa committed:
Merge pull request #967 from wolftankk/develop

Azubu support

chrippa committed:
Merge pull request #968 from chvrn/plugin

Plugin for Expressen

chrippa committed:
Merge pull request #970 from intact/douyutv-fix

Updated douyutv.com plugin

chrippa committed:
Merge pull request #971 from intact/add-streamupcom

Add streamupcom

chrippa committed:
Merge pull request #978 from Ell/add-livecodingtv-plugin

Added support for livecoding.tv

Commit:
Support for Tour de France stream

blxd committed:
add user agent header to the tvcatchup plugin

Commit:
add support to afreecatv.com.tw

WeinerRinkler committed:
First version

WeinerRinkler committed:
Error fixed when streamer offline or invalid

boekkooi committed:
NPO token fix

intact committed:
plugins.streamupcom: Update for API change

asermax committed:
plugins.crunchyroll: added support for locale selection

jtsymon committed:
Sort list of streams by quality

asermax committed:
plugins.crunchyroll: use locale parameter on the header's user-agent as well

chrippa committed:
Merge pull request #988 from michie1/nos-tdf

Support for Tour de France stream

chrippa committed:
Merge pull request #991 from blxd/bugfix/tvcatchup

TVCatchup.com bugfix

chrippa committed:
Merge pull request #995 from Summon528/patch-1

add support to afreecatv.com.tw

chrippa committed:
Merge pull request #1013 from boekkooi/issue-921

NPO token fix

chrippa committed:
Merge pull request #1007 from WeinerRinkler/develop

add support for younow.com

chrippa committed:
Merge pull request #1017 from intact/fix-streamupcom

plugins.streamupcom: Update for API change

chrippa committed:
Merge pull request #1018 from asermax/crunchyroll-select-locale

plugins.crunchyroll: added support for locale selection

chrippa committed:
cli: Close output on exit.

Commit:
plugins.afreeca: Fix HLS stream.

Resolves #982.

chrippa committed:
Show a brief usage when no option is specified.

Resolves #1009.

chrippa committed:
cli: Fix typo.

jtsymon committed:
Avoid sorting streams twice

chrippa committed:
Merge pull request #1019 from jtsymon/develop

Sort list of streams by quality

chrippa commented:
Not sure I understand your request, but if you want to skip a segment and move to the next one if it is too slow to download you can do it by setting the timeout lower and the maximum attempts to 1.

HLS example, this will move on to the next segment if it failed to download a segment under 4 seconds:

    livestreamer --hls-segment-attempts=1 --hls-segment-timeout=4

todostreaming commented:
Yes, I guess I have not explained correctly. I mean, if there is an internet disruption and no segments are accessible during a period of time (say 1 or 2 minutes), just recycle the old segments in the output, just repeating the latest ones. Could be an option of: "--loop-hs-segments 3", that would repeat the latest 3 segments until new segments are again accessible from the HLS server

chrippa commented:
I see, I don't think that's really possible though. Segments must be played in the correct order or the player will run into playback issues, such as audio sync issues, freezing video or just stop completly. 

The only way to play a segment from the past correctly would be to reinitialize the video decoder, but that is something we can't do from our end since we just feed the player a stream of data.

todostreaming commented:
I have tested disordering .ts segments in an m3u8 using ffplay as player, and it plays smoothly. I dont know other players behavior, however it would be only an option for those who want to use it.
Thanks for your replies.

pulviscriptor committed:
GoodGame URL parse fix

chrippa committed:
Merge pull request #1030 from pulviscriptor/develop

GoodGame URL parse fix

charmander committed:
plugins.picarto: Update for API and URL change

Feverqwe committed:
goodgame ddos validation

matiaspl committed:
Update periscope.py

Fix periscope url according to currently used schema.

sulami committed:
azubutv: set video_player to None if stream is offline

This prevents an error with traceback and instead just prints "No
streams found" like it should if the stream in question is not live.

steven7851 committed:
plugins.douyutv: Use new api.

Use douyutv new api for android with HLS stream feature.

neutric committed:
Update issues.rst

Fix spelling

meise committed:
plugin: added media_ccc_de api and protocol changes

* the media.ccc.de url scheme was changed after upgrading to a new frontend
* streaming.media.ccc.de supports also https

meise committed:
docs/plugin_matrix: removed needless characters

paulguy committed:
Maybe fixed ustream validation failure.

paulguy committed:
More strict test for weird stream.

intact committed:
plugins.dailymotion: Add HLS streams support

Commit:
Now catching socket errors on send

Commit:
Allow https urls for nrk.no.

pavlos256 committed:
Add antenna.gr plugin

pavlos256 committed:
Update plugin matrix for antenna

sbstp committed:
update the streamup.com plugin

Feverqwe committed:
add stream_id with words

Thx @flijloku

ph0o committed:
Create servustv.py

javiercantero committed:
Add new parameter to Twitch usher URL

Livestreamer was getting a "Player must be spectre-aware" error when it
tried to get the list of available HLS streams from some Twitch channels
(those playing a playlist?). To avoid this error, the usher service URL
now requires a new parameter called "allow_spectre" to be defined. Since
I don't know how this "spectre" thing works, I set the parameter to
"false".

intact committed:
plugins.npo: Fix Python 3 compatibility

chrippa committed:
Merge pull request #1192 from intact/fix-npo

plugins.npo: Fix Python 3 compatibility

chrippa committed:
Merge pull request #1174 from javiercantero/twitch-spectre

Add new parameter to Twitch usher URL

chrippa committed:
Merge pull request #1077 from Feverqwe/develop

goodgame ddos validation

chrippa committed:
Merge pull request #1148 from sbstp/streamup

Update the streamup.com plugin

chrippa committed:
Merge pull request #1135 from pavlos256/new-plugin-antenna

New plugin: antenna

chrippa committed:
Merge pull request #1133 from oyvindln/nrk-https

Allow https urls for nrk.no.

chrippa committed:
Merge pull request #1130 from Globidev/fix-crash-on-httpserver-sigpipe

Fix crashes on HTTPServer's SIGPIPE

chrippa committed:
Merge pull request #1129 from intact/fix-dailymotion

plugins.dailymotion: Add HLS streams support

chrippa committed:
Merge pull request #1126 from paulguy/ustream-fix

Possible fix for ustream

chrippa committed:
Merge pull request #1116 from meise/media_ccc_de_api_change

plugin: added media_ccc_de api and protocol changes

chrippa committed:
Merge pull request #1115 from meise/plugin_matrix_typo

docs/plugin_matrix: removed needless characters

chrippa committed:
Merge pull request #1106 from steven7851/patch-2

plugins.douyutv: Use new api.

chrippa committed:
Merge pull request #1092 from sulami/develop

azubutv: set video_player to None if stream is offline

chrippa committed:
Merge pull request #1076 from charmander/fix-picarto

plugins.picarto: Update for API and URL change

chrippa committed:
Merge branch 'pr/1168' into develop

chrippa committed:
Merge branch 'pr/1108' into develop

chrippa committed:
Merge branch 'pr/1089' into develop

chrippa committed:
travis: Use new artifacts tool.

chrippa committed:
docs: Fix readthedocs build.

chrippa committed:
travis: Build installer exe aswell.
